test/java/integration/AndroidEngineSocketTagTest.java

```diff
@@ -0,0 +1,392 @@
+package test.kotlin.integration;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import android.content.Context;
+import androidx.test.core.app.ApplicationProvider;
+import io.envoyproxy.envoymobile.AndroidEngineBuilder;
+import io.envoyproxy.envoymobile.Engine;
+import io.envoyproxy.envoymobile.EnvoyError;
+import io.envoyproxy.envoymobile.FinalStreamIntel;
+import io.envoyproxy.envoymobile.LogLevel;
+import io.envoyproxy.envoymobile.RequestHeaders;
+import io.envoyproxy.envoymobile.RequestHeadersBuilder;
+import io.envoyproxy.envoymobile.RequestMethod;
+import io.envoyproxy.envoymobile.ResponseHeaders;
+import io.envoyproxy.envoymobile.ResponseTrailers;
+import io.envoyproxy.envoymobile.Stream;
+import io.envoyproxy.envoymobile.StreamIntel;
+import io.envoyproxy.envoymobile.UpstreamHttpProtocol;
+import io.envoyproxy.envoymobile.engine.AndroidJniLibrary;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.util.AbstractMap.SimpleImmutableEntry;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import okhttp3.mockwebserver.Dispatcher;
+import okhttp3.mockwebserver.MockResponse;
+import okhttp3.mockwebserver.MockWebServer;
+import okhttp3.mockwebserver.RecordedRequest;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.robolectric.RobolectricTestRunner;
+
+@RunWith(RobolectricTestRunner.class)
+public class AndroidEngineSocketTagTest {
+
+  private final MockWebServer mockWebServer = new MockWebServer();
+  private Engine engine;
+
+  @BeforeClass
+  public static void loadJniLibrary() {
+    AndroidJniLibrary.loadTestLibrary();
+  }
+
+  @Before
+  public void setUpEngine() throws Exception {
```

Comment:
wow this seems like a lot of boilerplate to send a request and check for an error response. Maybe @jpsim or @goaway can comment if there's any low hanging fruit for clean up, or add "refactor for reuse" to the ever growing list of things we should do?

Reply:
I like the way you think! It looks like the `Response` and `RequestScenario` classes are used in a few different tests. How 'bout this:

In parallel with this PR (so as to avoid blocking it) I refactor those tests to move `Response` and `RequestScenario` into stand-alone helper classes, and depending on which PR lands first, make sure that after both land, the tests in this file use the new helper?

Reply:
https://github.com/envoyproxy/envoy-mobile/pull/2444 landed so you should be able to benefit from it in your PR.